File: weekend-rush/backend/api/customers.py
from flask import Blueprint, jsonify, request

from api import token_required
from models import Customer, db

customers_bp = Blueprint("customers", __name__)


@customers_bp.get("/")
@token_required
def list_customers():
    rows = Customer.query.order_by(Customer.name).all()
    return jsonify({"customers": [row.to_dict() for row in rows]}), 200


@customers_bp.post("/")
@token_required
def create_customer():
    data = request.get_json(silent=True) or {}
    name = (data.get("name") or "").strip()
    phone = (data.get("phone") or "").strip()
    nfc_uid = (data.get("nfc_uid") or "").strip() or None

    if not name:
        return jsonify({"error": "Name is required"}), 400
    if not phone.isdigit() or len(phone) < 10:
        return jsonify({"error": "Phone must be at least 10 digits"}), 400
    if Customer.query.filter_by(phone=phone).first():
        return jsonify({"error": "Phone already registered"}), 409
    if nfc_uid and Customer.query.filter_by(nfc_uid=nfc_uid).first():
        return jsonify({"error": "NFC UID already assigned"}), 409

    customer = Customer(
        name=name,
        phone=phone,
        is_member=bool(data.get("is_member", False)),
        nfc_uid=nfc_uid,
    )
    db.session.add(customer)
    db.session.commit()
    return jsonify({"message": "Customer added", "customer": customer.to_dict()}), 201


@customers_bp.put("/<int:customer_id>")
@token_required
def update_customer(customer_id):
    customer = Customer.query.get(customer_id)
    if customer is None:
        return jsonify({"error": "Customer not found"}), 404

    data = request.get_json(silent=True) or {}
    if "name" in data:
        name = (data.get("name") or "").strip()
        if not name:
            return jsonify({"error": "Name cannot be empty"}), 400
        customer.name = name
    if "phone" in data:
        phone = (data.get("phone") or "").strip()
        if not phone.isdigit() or len(phone) < 10:
            return jsonify({"error": "Phone must be at least 10 digits"}), 400
        existing = Customer.query.filter_by(phone=phone).first()
        if existing and existing.id != customer.id:
            return jsonify({"error": "Phone already registered"}), 409
        customer.phone = phone
    if "is_member" in data:
        customer.is_member = bool(data.get("is_member"))
    if "nfc_uid" in data:
        nfc_uid = (data.get("nfc_uid") or "").strip() or None
        if nfc_uid:
            existing = Customer.query.filter_by(nfc_uid=nfc_uid).first()
            if existing and existing.id != customer.id:
                return jsonify({"error": "NFC UID already assigned"}), 409
        customer.nfc_uid = nfc_uid

    db.session.commit()
    return jsonify({"message": "Customer updated", "customer": customer.to_dict()}), 200


@customers_bp.post("/<int:customer_id>/toggle-member")
@token_required
def toggle_member(customer_id):
    customer = Customer.query.get(customer_id)
    if customer is None:
        return jsonify({"error": "Customer not found"}), 404
    customer.is_member = not customer.is_member
    db.session.commit()
    return jsonify({"message": "Membership updated", "customer": customer.to_dict()}), 200


@customers_bp.post("/<int:customer_id>/nfc")
@token_required
def update_nfc(customer_id):
    customer = Customer.query.get(customer_id)
    if customer is None:
        return jsonify({"error": "Customer not found"}), 404

    data = request.get_json(silent=True) or {}
    nfc_uid = (data.get("nfc_uid") or "").strip() or None
    if nfc_uid:
        existing = Customer.query.filter_by(nfc_uid=nfc_uid).first()
        if existing and existing.id != customer.id:
            return jsonify({"error": "NFC UID already assigned"}), 409
    customer.nfc_uid = nfc_uid
    db.session.commit()
    return jsonify({"message": "NFC UID updated", "customer": customer.to_dict()}), 200


@customers_bp.delete("/<int:customer_id>")
@token_required
def delete_customer(customer_id):
    customer = Customer.query.get(customer_id)
    if customer is None:
        return jsonify({"error": "Customer not found"}), 404
    db.session.delete(customer)
    db.session.commit()
    return jsonify({"message": "Customer removed"}), 200

File: weekend-rush/frontend/js/customers.js
/* Customer directory: membership and NFC management. */
document.addEventListener("DOMContentLoaded", () => {
  if (!initStaffShell()) return;
  document.getElementById("customer-form").addEventListener("submit", addCustomer);
  loadCustomers();
});

async function loadCustomers() {
  const message = document.getElementById("list-msg");
  const body = document.getElementById("customers-body");
  try {
    const data = await API.getCustomers();
    if (!data.customers.length) {
      body.innerHTML = '<tr><td colspan="5" class="muted">No customers yet.</td></tr>';
      return;
    }

    body.innerHTML = data.customers
      .map(
        (customer) => `
      <tr>
        <td>${customer.name}</td>
        <td>${customer.phone}</td>
        <td>${customer.is_member ? '<b class="lime">Member</b>' : "Guest"}</td>
        <td><input id="nfc-${customer.id}" type="text" value="${customer.nfc_uid || ""}" /></td>
        <td>
          <div class="actions">
            <button class="btn sm" data-toggle="${customer.id}">Toggle member</button>
            <button class="btn sm dark" data-nfc="${customer.id}">Save NFC</button>
            <button class="btn sm danger admin-only" data-delete="${customer.id}">Delete</button>
          </div>
        </td>
      </tr>`
      )
      .join("");

    body.querySelectorAll("[data-toggle]").forEach((button) =>
      button.addEventListener("click", () =>
        runAction(() => API.toggleMember(Number(button.dataset.toggle)), "Membership updated.")
      )
    );
    body.querySelectorAll("[data-nfc]").forEach((button) =>
      button.addEventListener("click", () => {
        const id = Number(button.dataset.nfc);
        const value = document.getElementById(`nfc-${id}`).value.trim();
        runAction(() => API.setNfc(id, value), "NFC UID saved.");
      })
    );
    body.querySelectorAll("[data-delete]").forEach((button) =>
      button.addEventListener("click", () =>
        runAction(() => API.deleteCustomer(Number(button.dataset.delete)), "Customer removed.")
      )
    );

    lockForStaff(body);
    showMessage(message, "", "info");
  } catch (error) {
    showMessage(message, error.message, "error");
  }
}

async function runAction(action, successText) {
  const message = document.getElementById("list-msg");
  try {
    await action();
    showMessage(message, successText, "success");
    loadCustomers();
  } catch (error) {
    showMessage(message, error.message, "error");
  }
}

async function addCustomer(event) {
  event.preventDefault();
  const message = document.getElementById("customer-msg");
  const name = document.getElementById("cust-name").value.trim();
  const phone = document.getElementById("cust-phone").value.trim();
  const nfc = document.getElementById("cust-nfc").value.trim();

  if (!name) {
    showMessage(message, "Name is required.", "error");
    return;
  }
  if (!/^[0-9]{10,}$/.test(phone)) {
    showMessage(message, "Phone must be at least 10 digits.", "error");
    return;
  }

  try {
    await API.addCustomer({
      name,
      phone,
      nfc_uid: nfc,
      is_member: document.getElementById("cust-member").checked,
    });
    showMessage(message, "Customer added.", "success");
    document.getElementById("customer-form").reset();
    loadCustomers();
  } catch (error) {
    showMessage(message, error.message, "error");
  }
}
